segmentstore/server/src/main/java/io/pravega/segmentstore/server/CacheManager.java

```diff
@@ -415,6 +430,10 @@ private boolean adjustOldestGeneration(CacheStatus currentStatus) {
         return isAdjusted;
     }
 
+    private void adjustNonEssentialEnabled() {
+        this.essentialEntriesOnly.set(this.lastCacheState.get().getUsedBytes() >= this.policy.getCriticalThreshold());
```

Comment:
@RaulGracia You are correct that that's how a cache should work. A LRU cache would work that way. However implementing such a scheme here is prohibitive; we have many cache clients that store their data and each of those clients are maintained by different segment containers.

As such, I put the "cache entry generation" mechanism in place as a proxy for LRU order. When the cache is full, we do attempt to clean up old entries that are not essential. That works well except in cases of very high congestion (lots of writes and lots of reads at the same time).

We only increment the generation once every 1s, or every time we try to perform a cache eviction. Due to this needing to be sequential (can't have multiple concurrent eviction processes at the same time), there is a "floor" on the minimum duration of each such generation (i.e., single or double-digit milliseconds). If we have high congestion, it is likely that the cache manager does not have sufficient granularity in the entries to locate eligible entries to evict (all non-essential entries are written in the last generation). At this point the cache manager begins thrashing and it uses more and more resources of the system, leading to it running continuously. When it does this, unless the external congestion stops, it will stop providing more benefit to the system. 

Switching the cache system to a LRU (using a true Linked List or whatever) would be resource prohibitive and CPU prohibitive, so it's unlikely we'll want to head in that direction. 

As such, one of our remaining options is to help reduce congestion. It feels reasonable to me that if the cache is at or above 90% (remember that the Cache Manager aims to keep it below 75%), then we are already "in trouble". At this point, we can eliminate non-essential traffic to alleviate congestion. We want to let the Cache Manager do its job, while also leaving that remaining 10% (or less) for system-critical data (metadata updates).